packages/core/src/sdk/product/usePageProductsQuery.ts

```diff
@@ -0,0 +1,138 @@
+import { gql } from '@faststore/graphql-utils'
+import { useSearch } from '@faststore/sdk'
+import {
+  ClientProductsQueryQuery,
+  ClientProductsQueryQueryVariables,
+} from '@generated/graphql'
+import { MutableRefObject, useEffect, useRef, useCallback } from 'react'
+import type { QueryOptions } from '../graphql/useQuery'
+import { useQuery } from 'src/sdk/graphql/useQuery'
+import { useSWRConfig } from 'swr'
+import { prefetchQuery } from '../graphql/prefetchQuery'
+import { useLocalizedVariables } from './useLocalizedVariables'
+import { ProductsPerPage } from '../overrides/PageProvider'
+
+export const query = gql`
+  query ClientProductsQuery(
+    $first: Int!
+    $after: String
+    $sort: StoreSort!
+    $term: String!
+    $selectedFacets: [IStoreSelectedFacet!]!
+  ) {
+    ...ClientProducts
+    search(
+      first: $first
+      after: $after
+      sort: $sort
+      term: $term
+      selectedFacets: $selectedFacets
+    ) {
+      products {
+        pageInfo {
+          totalCount
+        }
+        edges {
+          node {
+            ...ProductSummary_product
+          }
+        }
+      }
+    }
+  }
+`
+
+export const useProductsQueryPrefetch = (
+  variables: ClientProductsQueryQueryVariables,
+  options?: QueryOptions
+) => {
+  const localizedVariables = useLocalizedVariables(variables)
+  const { cache } = useSWRConfig()
+
+  return useCallback(
+    () => prefetchQuery(query, localizedVariables, { cache, ...options }),
+    [localizedVariables, cache, options]
+  )
+}
+
+export const useProductsPrefetch = (page: number | null) => {
+  const {
+    itemsPerPage,
+    state: { sort, term, selectedFacets },
+  } = useSearch()
+
+  const prefetch = useProductsQueryPrefetch({
+    first: itemsPerPage,
+    after: (itemsPerPage * (page ?? 0)).toString(),
+    sort,
+    term: term ?? '',
+    selectedFacets,
+  })
+
+  useEffect(() => {
+    if (page !== null) {
+      prefetch()
+    }
+  }, [page, prefetch])
+}
+
+function updatesProductsPerPageRef(
+  productsPerPageRef: MutableRefObject<ProductsPerPage[]>,
+  page: number,
+  data: ClientProductsQueryQuery
+) {
+  const productsPerPage = productsPerPageRef.current
+  const currentProductsPerPage = { page, data }
+
+  if (productsPerPage.length === 0) {
+    productsPerPage.push(currentProductsPerPage)
+    return { productsPerPage, currentProductsPerPage }
+  }
+
+  const index = productsPerPage.findIndex((item) => item.page === page)
+  const lastItem = productsPerPage[productsPerPage.length - 1]
+  const shouldReplaceItem = index !== -1
+  const shouldAddNext = page > lastItem.page
+
+  if (shouldReplaceItem) {
+    productsPerPage.splice(index, 1, currentProductsPerPage)
+  } else if (shouldAddNext) {
+    productsPerPage.push(currentProductsPerPage)
+  } else {
+    productsPerPage.unshift(currentProductsPerPage)
+  }
+  return { productsPerPage, currentProductsPerPage }
+}
```

Comment:
I think we should always return a new array instance, avoiding mutating the original array altogether. This is a good practice in the JS world and avoids errors that may be caused by shared array references.

I think we could also remove `useRef` by doing this.

Besides all that, I think this function can be simpler. If `page < productsPerPage[0]`, add to the beginning, else, add to the end.

Reply:
Well, I agree with you. 
I simplify this logic [here](https://github.com/vtex/faststore/pull/1967/commits/e09ae56d029f53ccadc759d3448e1d3b1356e8e7) Also now is returning a new array instance without `useRef`, but  we need to get the `productsPerPage` state from context in `ProductGaleryPage` and pass to the hook so that the things keep working as expected.

Reply:
I've commented about it in my last review